packages/dev/core/src/Rendering/IBLShadows/iblShadowsSpatialBlurPass.ts

```diff
@@ -106,67 +114,105 @@ export class _IblShadowsSpatialBlurPass {
-            extraInitializations: (useWebGPU: boolean, list: Promise<any>[]) => {
-                if (useWebGPU) {
-                    list.push(import("../../ShadersWGSL/iblShadowSpatialBlur.fragment"));
+            extraInitializationsAsync: async () => {
+                if (isWebGPU) {
+                    await Promise.all([import("../../ShadersWGSL/iblShadowSpatialBlur.fragment")]);
                 } else {
-                    list.push(import("../../Shaders/iblShadowSpatialBlur.fragment"));
+                    await Promise.all([import("../../Shaders/iblShadowSpatialBlur.fragment")]);
                 }
             },
         };
-        this._outputPP = new PostProcess("spacialBlurPP", "iblShadowSpatialBlur", ppOptions);
-        this._outputPP.autoClear = false;
-        this._outputPP.onApplyObservable.add((effect) => {
-            this._updatePostProcess(effect);
+        this._outputTexture = new ProceduralTexture(
+            "spatialBlurPass",
+            {
+                width: this._engine.getRenderWidth(),
+                height: this._engine.getRenderHeight(),
+            },
+            "iblShadowSpatialBlur",
+            this._scene,
+            textureOptions,
+            false,
+            false,
+            Constants.TEXTURETYPE_UNSIGNED_INT
+        );
+        this._outputTexture.refreshRate = -1;
+        this._outputTexture.autoClear = false;
+
+        // Need to set all the textures first so that the effect gets created with the proper uniforms.
+        this._update();
+
+        let counter = 0;
+        this._scene.onBeforeRenderObservable.add(() => {
+            counter = 0;
+        });
+        this._scene.onAfterRenderTargetsRenderObservable.add(() => {
+            if (++counter == 2) {
+                if (this.enabled && this._outputTexture.isReady()) {
+                    this._update();
+                    this._outputTexture.render();
+                }
+            }
         });
     }
 
-    private _updatePostProcess(effect: Effect) {
+    private _update() {
+        this._outputTexture.setTexture("voxelTracingSampler", this._renderPipeline.getVoxelTracingTexture());
         const iterationCount = 1;
-        effect.setVector4("blurParameters", new Vector4(iterationCount, this._worldScale, 0.0, 0.0));
-        const prePassRenderer = this._scene.prePassRenderer;
-        if (prePassRenderer) {
-            const wnormalIndex = prePassRenderer.getIndex(Constants.PREPASS_WORLD_NORMAL_TEXTURE_TYPE);
-            const depthIndex = prePassRenderer.getIndex(Constants.PREPASS_DEPTH_TEXTURE_TYPE);
-            if (wnormalIndex >= 0) effect.setTexture("worldNormalSampler", prePassRenderer.getRenderTarget().textures[wnormalIndex]);
-            if (depthIndex >= 0) effect.setTexture("linearDepthSampler", prePassRenderer.getRenderTarget().textures[depthIndex]);
+        this._outputTexture.setVector4("blurParameters", new Vector4(iterationCount, this._worldScale, 0.0, 0.0));

```

Comment:
Same as earlier, try to avoid creating new objects each frame, for GC purpose.

Reply:
Done